sw/otbn/crypto/sha3-shake.s

```diff
@@ -0,0 +1,753 @@
+  /* Mask out the lowest 2 bits of the address for the exact offset */
+  andi x29, x5, 0x3
+  slli x29, x29, 3
+  addi x30, x0, 0x1F
+  sll  x30, x30, x29
+  xor  x28, x28, x30
+  sw   x28, 0(x6)
+
+  /* 4B-aligned load st[rsiz-1] */
+  lw   x5, 204(x10)
+  addi x5, x5, -1 
+  add  x5, x10, x5
+  and  x6, x5, x12
+  lw   x28, 0(x6)
+  /* Mask out the lowest 2 bits of the address for the exact offset */
+  andi x29, x5, 0x3
+  slli x29, x29, 3 
+  addi x30, x0, 0x80
+  sll  x30, x30, x29
+  xor  x28, x28, x30
+  sw   x28, 0(x6)
+
+  jal  x1, keccakf
+  sw   x0, 200(x10)
+  ret
+
+/**
+ * shake_out
+ *
+ * Get SHAKE output
+ *
+ * @param[in]    x12: desired output length
+ * @param[in]    x11: output pointer
+ * @param[inout] x10: pointer to 212 bytes of memory for holding the context.
+                   32B aligned.
+ *
+ * clobbered registers: x5 to x7, x13 to x17, x28 to x31, 
+ *                      w0 to w13, w21 to w30
+ * clobbered flag groups: FG0 
+ */
+.global shake_out
+shake_out:
+  /* Load pt as j */
+  lw   x5, 200(x10)
+  /* Load constant for address alignment */
+  li   x30, 0xFFFFFFFC
+  li   x17, 0
+  addi x17, x17, -1
+  /* Data counter */
+  li   x13, 0
+
+  beq  x12, x0, _shake_out_skip_loop
+  LOOP x12, 27
+    /* Load rsiz */
+    lw   x29, 204(x10)
+    sub  x28, x5, x29
+    srli x28, x28, 31
+    bne  x28, x0, _shake_out_skip
+
+    jal  x1, keccakf
+    xor  x5, x5, x5 /* j = 0 */
+
+_shake_out_skip:
+    /* 4B-aligned load from source address */
+    add  x7, x5, x10
+    addi x5, x5, 1
+    and  x28, x7, x30
+    lw   x28, 0(x28)
+    /* Mask out the lowest 2 bits of the address for the exact offset */
+    andi x7, x7, 0x3
+    slli x7, x7, 3 
+    srl  x28, x28, x7
+    andi x28, x28, 0xFF
+
+    /* 4B-aligned load from destination address */
+    add  x7, x11, x13
+    addi x13, x13, 1
+    and  x16, x7, x30
+    lw   x29, 0(x16)
+    /* Mask out the lowest 2 bits of the address for the exact offset */
+    andi x14, x7, 0x3
+    slli x14, x14, 3
+    /* Clear index for insertion */
+    addi x31, x0, 0xFF
+    sll  x15, x31, x14
+    xor  x15, x17, x15
+    and  x29, x29, x15
+
+    /* Insert */ 
+    sll  x28, x28, x14
+    or   x29, x29, x28
+
+    /* Aligned store to output */
+    sw   x29, 0(x16)
+
+_shake_out_skip_loop:
+  /* Store j as pt */
+  sw x5, 200(x10)
+  ret
+
+/**
+ * Compute Keccak-f permutation
+ *
+ * @param[in] x10: Pointer to context 
+ * 
+ * clobbered registers: x5 to x6, x28 to x29, x31, 
+ *                      w0 to w13, w21 to w30
+ * clobbered flag groups: FG0
+ *
+ * Upercase L_i denotes the i-th lane from the state array.
+ * ^ denotes exclusive-OR 
+ * & denotes logical AND 
+ * << denotes left shift
+ */
+
+.global keccakf
+keccakf:
+  /* Copy context pointer */
+  add     x5, x0, x10 
+
+  /* Load necessary constants:
+     w11 <= (1 <<  64) - 1
+     w12 <= (1 << 256) - 1
+     w13 <= mask_top_1 */
+  li      x29, 12
+  la      x31, mask_bn 
+  bn.lid  x29, 0(x31++) 
```

Comment:
nit: If you just need an all-ones value, you can get it faster and without needing to store a constant using the all-zero register:
```suggestion
  bn.not  x29, w31 
```